src/Databases/DatabaseOnDisk.cpp

```diff
@@ -324,31 +325,36 @@ void DatabaseOnDisk::dropTable(ContextPtr local_context, const String & table_na
 
     StoragePtr table = detachTable(local_context, table_name);
 
-    /// This is possible for Lazy database.
-    if (!table)
-        return;
-
     bool renamed = false;
     try
     {
         fs::rename(table_metadata_path, table_metadata_path_drop);
         renamed = true;
-        table->drop();
-        table->is_dropped = true;
-
-        fs::path table_data_dir(local_context->getPath() + table_data_path_relative);
-        if (fs::exists(table_data_dir))
-            (void)fs::remove_all(table_data_dir);
+        // The table might be not loaded for Lazy database engine.
+        if (table)
```

Comment:
I don't see how is it possible. We will throw exception https://github.com/ClickHouse/ClickHouse/blob/8f5c8fdb27c2528b3ff4d0bad31b7047ec690949/src/Databases/DatabaseLazy.cpp#L197-L198.

Reply:
As I understood:
https://github.com/ClickHouse/ClickHouse/blob/8f5c8fdb27c2528b3ff4d0bad31b7047ec690949/src/Databases/DatabaseLazy.cpp#L192-L205

If the table was recently touched it will be present in `tables_cache[name].table`, if not here will be nullptr.
So the return value from  Detach function is `it->second.table` not just `it`.

Here we are loading the table if `tables_cache[name].table` is a null: https://github.com/ClickHouse/ClickHouse/blob/8f5c8fdb27c2528b3ff4d0bad31b7047ec690949/src/Databases/DatabaseLazy.cpp#L133

Here we are doing reset if lifetime is expired:
https://github.com/ClickHouse/ClickHouse/blob/8f5c8fdb27c2528b3ff4d0bad31b7047ec690949/src/Databases/DatabaseLazy.cpp#L284 
https://github.com/ClickHouse/ClickHouse/blob/8f5c8fdb27c2528b3ff4d0bad31b7047ec690949/src/Databases/DatabaseLazy.cpp#L306-L307


Summary: `table` might be empty if it has not been touched recently